### ISSUE TITLE
With baseurl=http://prerelease.keybase.io/rpm/x86_64/, media.1/media should 404?

### ISSUE THREAD

**Issue by Espionage724** [keybase/keybase-issues#2987]
I'm trying to add http://prerelease.keybase.io/rpm/x86_64 as a repo on openSUSE but it wants to try probing http://prerelease.keybase.io/rpm/x86_64/media.1/media as well. If I understand right, if the media.1/media doesn't exist, it ignores that and then uses the baseurl. But in this case, it's denying access, which causes zypper to error-out instead.

Related openSUSE bug report: https://bugzilla.opensuse.org/show_bug.cgi?id=1032348

This post here claims that having the media.1/media URL 404 instead of 403'ing works: https://forums.insynchq.com/t/repository-for-opensuse/954/60

**Comment by maxtaco** [comment]
cc @cjb and @oconnor663

**Comment by cjb** [comment]
Thanks for the report.  It's pretty frustrating that the upstream bug is closed wontfix.  I hope it's okay if I CC @aspiers on the chance that he's interested in the experience.

My intuition is that servers returning 403 instead of 404 are properly configured -- they're choosing not to leak metadata about whether a path is valid to a user which doesn't have permissions on that path.

**Comment by aspiers** [comment]
Thanks for the cc, unfortunately I'm swamped. @cseader submitted the upstream bug, maybe he can help out here.

**Comment by oconnor663** [comment]
What's the simplest workaround for this? Can we create an empty file at that path and call it a day? Or would we need to special-case this path in S3 configs?

**Comment by AmandaCameron** [comment]
@cjb Perhaps I'm wrong, but what exact "metadata" could be leaked from a S3 bucket used just for serving apt / rpm / etc folders? zypper / apt / etc don't tend to support passing S3 credentials through to the backing repositories, so it's not like per-user differences make much sense in this context ( Unless I'm missing something, here )

**Comment by cjb** [comment]
@AmandaCameron You're not wrong, I'm just speaking generally about the decision of web server software, without any knowledge of what users are serving with it, choosing to return 403 instead of 404 for non-existent paths the user has no permissions on.

Because web server software can reasonably make this decision, I think it's a mistake to interpret 403 as confirming the existence of the file, as Zypper apparently does.

**Comment by AmandaCameron** [comment]
Then I guess my question in this area goes to @oconnor663 on whether it'd be acceptable to change the bucket's settings to spit out a 404 instead of a 403 as the default.

**Comment by cjb** [comment]
It's possible for us to do this, but it's frustrating that Keybase has to special case its web server configuration to make OpenSUSE happy, and it's presumably also frustrating that OpenSUSE has to reach out to every RPM publisher hosting on S3 and ask each of them to do this, and it seems like everyone could save a lot of time if OpenSUSE reconsidered the semantics of this 403, and I guess I was hoping that if we ignored this for a while someone would realize it was a bad idea and fix it in Zypper :)

**Comment by oconnor663** [comment]
I've added public list permissions for the S3 bucket that backs this URL. I think that should fix the issue. Please let me know if you still see it.

**Comment by cjb** [comment]
> frustrating that OpenSUSE has to reach out to every RPM publisher hosting on S3 and ask each of them to do this, and it seems like everyone could save a lot of time if OpenSUSE reconsidered the semantics of this 403, and I guess I was hoping that if we ignored this for a while someone would realize it was a bad idea

They didn't :(

**Comment by oconnor663** [comment]
Next time I have to package a desktop app I'm just going to distribute it as a VirtualBox OVA :-D

**Comment by AmandaCameron** [comment]
Great news, @oconnor663! Hopefully https://github.com/keybase/client/pull/5226 (or something liek it) can land next, because otherwise we have to pat zypper on the back every update to tell it it's okay.

Thanks for all the hard work y'all do, it's definately appreciated. :)
